src/trace_ng/log.zig

```diff
@@ -0,0 +1,600 @@
+    const StanardErrLogger = struct {
+        const Self = @This();
+        max_level: Level,
+        exit_sig: *std.atomic.Value(bool),
+        allocator: Allocator,
+        recycle_fba: RecycleFBA,
+        max_buffer: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) Self {
+            return .{
+                .allocator = config.allocator,
+                .recycle_fba = RecycleFBA.init(config.allocator, config.max_buffer) catch @panic("could not create RecycleFBA"),
+                .max_buffer = config.max_buffer,
+                .max_level = config.max_level,
+                .exit_sig = config.exit_sig,
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = null,
+            };
+        }
+
+        fn unscoped(self: *Self) *UnscopedLogger {
+            return @ptrCast(self);
+        }
+
+        fn withScope(self: *Self, comptime new_scope: anytype) *StandardLogger(new_scope) {
+            return @ptrCast(self);
+        }
+
+        pub fn spawn(self: *Self) void {
+            self.handle = std.Thread.spawn(.{}, Self.run, .{self}) catch @panic("could not spawn Logger");
+        }
+
+        pub fn deinit(self: *Self) void {
+            if (self.handle) |*handle| {
+                self.exit_sig.store(true, .seq_cst);
+                handle.join();
+            }
+            self.channel.close();
+            self.channel.deinit();
+            self.recycle_fba.deinit();
+        }
+
+        pub fn run(self: *Self) void {
+            while (!self.exit_sig.load(.seq_cst)) {
+                std.time.sleep(std.time.ns_per_ms * 5);
+                const messages = self.channel.drain() orelse {
+                    // channel is closed
+                    return;
+                };
+                defer self.channel.allocator.free(messages);
+
+                for (messages) |message| {
+                    const writer = std.io.getStdErr().writer();
+                    logfmt.writeLog(writer, message) catch @panic("logging failed");
+                }
+                self.recycle_fba.mux.lock();
+                self.recycle_fba.alloc_allocator.reset();
+                self.recycle_fba.mux.unlock();
+            }
+        }
+
+        pub fn log(self: *Self, level: Level, message: []const u8) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const logMessage = logfmt.createLogMessage(&self.recycle_fba, self.max_buffer, level, maybe_scope, message, null, null, null);
+            self.channel.send(logMessage) catch @panic("could not send to channel");
```

Comment:
`@panic` is used a lot in the codebase but not sure if this is idiomatic/recommended

Reply:
It's probably best not to have any panics in the logger. I would only explicitly panic in the code when the following conditions are met:
- the panicking branch can only be triggered if a bug has been introduced in the code
- the error is likely to corrupt the entire application and make it unusable or unsafe to use
- the error cannot be recovered from without editing the code and recompiling a new binary